File: Swift/BENClient.swift
// BENClient.swift
import Foundation
import UIKit

public enum BEN2Error: Error, LocalizedError {
    case invalidResponse(Int, String)
    case notImage
    case encodeFailed
    public var errorDescription: String? {
        switch self {
        case .invalidResponse(let code, let body): return "HTTP \(code): \(body)"
        case .notImage: return "Server returned non-image data"
        case .encodeFailed: return "Could not encode image"
        }
    }
}

public final class BEN2Client {
    public static let shared = BEN2Client(
        apiKey: Bundle.main.object(forInfoDictionaryKey: "BEN2_API_KEY") as? String ?? "YOUR_API_KEY"
    )

    private let apiKey: String
    private let baseURL = URL(string: "https://api.backgrounderase.net/v2")!
    private let session: URLSession

    public init(apiKey: String, session: URLSession = .shared) {
        self.apiKey = apiKey
        self.session = session
    }

    // MARK: Public convenience (UIImage -> PNG bytes)
    public func removeBackground(image uiImage: UIImage,
                                 filename: String = "image.jpg",
                                 jpegQuality: CGFloat = 0.9,
                                 timeout: TimeInterval = 60) async throws -> Data {
        guard let data = uiImage.jpegData(compressionQuality: jpegQuality) else { throw BEN2Error.encodeFailed }
        return try await removeBackground(imageData: data, filename: filename, timeout: timeout)
    }

    // MARK: Core uploader (arbitrary Data)
    public func removeBackground(imageData: Data,
                                 filename: String,
                                 timeout: TimeInterval = 60) async throws -> Data {

        var request = URLRequest(url: baseURL)
        request.httpMethod = "POST"
        request.timeoutInterval = timeout

        let boundary = "----BEN2-\(UUID().uuidString)"
        request.setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")
        request.setValue(apiKey, forHTTPHeaderField: "x-api-key")

        let mime = Self.mimeType(for: filename)
        let body = Self.multipartBody(
            name: "image_file",
            filename: filename,
            mimeType: mime,
            fileData: imageData,
            boundary: boundary
        )
        request.httpBody = body
        request.setValue(String(body.count), forHTTPHeaderField: "Content-Length")

        let (data, response) = try await session.data(for: request)
        guard let http = response as? HTTPURLResponse else { throw URLError(.badServerResponse) }
        guard http.statusCode == 200 else {
            let msg = String(data: data, encoding: .utf8) ?? "<no body>"
            throw BEN2Error.invalidResponse(http.statusCode, msg)
        }
        return data // raw PNG bytes
    }

    // MARK: Helpers
    private static func mimeType(for filename: String) -> String {
        switch (filename as NSString).pathExtension.lowercased() {
        case "jpg", "jpeg": return "image/jpeg"
        case "png":         return "image/png"
        case "heic":        return "image/heic"
        case "webp":        return "image/webp"
        default:            return "application/octet-stream"
        }
    }

    private static func multipartBody(name: String,
                                      filename: String,
                                      mimeType: String,
                                      fileData: Data,
                                      boundary: String) -> Data {
        let CRLF = "\r\n"
        var data = Data()
        data.append("--\(boundary)\(CRLF)".data(using: .utf8)!)
        data.append("Content-Disposition: form-data; name=\"\(name)\"; filename=\"\(filename)\"\(CRLF)".data(using: .utf8)!)
        data.append("Content-Type: \(mimeType)\(CRLF)\(CRLF)".data(using: .utf8)!)
        data.append(fileData)
        data.append("\(CRLF)--\(boundary)--\(CRLF)".data(using: .utf8)!)
        return data
    }
}

File: CS/BENClient.cs
using System;
using System.IO;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

public static class BENClient
{
    private static readonly Uri Endpoint = new Uri("https://api.backgrounderase.net/v2");

    public static async Task<bool> RemoveBackgroundManualAsync(
        string src,
        string dst,
        string apiKey,
        CancellationToken ct = default)
    {
        var boundary = "----" + Guid.NewGuid().ToString("N");
        const string CRLF = "\r\n";

        var fileName = Path.GetFileName(src);
        var contentType = GuessMimeType(src);

        var header =
            $"--{boundary}{CRLF}" +
            $"Content-Disposition: form-data; name=\"image_file\"; filename=\"{fileName}\"{CRLF}" +
            $"Content-Type: {contentType}{CRLF}{CRLF}";

        var trailer = $"{CRLF}--{boundary}--{CRLF}";

        var fileBytes = await File.ReadAllBytesAsync(src, ct);
        var headerBytes = Encoding.ASCII.GetBytes(header);
        var trailerBytes = Encoding.ASCII.GetBytes(trailer);

        using var ms = new MemoryStream(headerBytes.Length + fileBytes.Length + trailerBytes.Length);
        ms.Write(headerBytes, 0, headerBytes.Length);
        ms.Write(fileBytes, 0, fileBytes.Length);
        ms.Write(trailerBytes, 0, trailerBytes.Length);

        using var content = new ByteArrayContent(ms.ToArray());
        content.Headers.ContentType = new MediaTypeHeaderValue("multipart/form-data");
        content.Headers.ContentType.Parameters.Add(new NameValueHeaderValue("boundary", boundary));

        using var req = new HttpRequestMessage(HttpMethod.Post, Endpoint) { Content = content };
        req.Headers.Add("x-api-key", apiKey);

        using var client = new HttpClient();
        using var resp = await client.SendAsync(req, HttpCompletionOption.ResponseHeadersRead, ct);
        var bytes = await resp.Content.ReadAsByteArrayAsync(ct);

        if (resp.IsSuccessStatusCode)
        {
            Directory.CreateDirectory(Path.GetDirectoryName(Path.GetFullPath(dst))!);
            await File.WriteAllBytesAsync(dst, bytes, ct);
            Console.WriteLine($"✅ Saved: {dst}");
            return true;
        }

        Console.Error.WriteLine($"❌ {(int)resp.StatusCode} {resp.ReasonPhrase}\n{TryDecodeUtf8(bytes)}");
        return false;
    }

    private static string GuessMimeType(string path)
    {
        var ext = Path.GetExtension(path).ToLowerInvariant();
        return ext switch
        {
            ".png"            => "image/png",
            ".jpg" or ".jpeg" => "image/jpeg",
            ".webp"           => "image/webp",
            ".gif"            => "image/gif",
            ".bmp"            => "image/bmp",
            ".tif" or ".tiff" => "image/tiff",
            ".heic"           => "image/heic",
            ".heif"           => "image/heif",
            _                 => "application/octet-stream"
        };
    }

    private static string TryDecodeUtf8(byte[] bytes)
    {
        try { return Encoding.UTF8.GetString(bytes); }
        catch { return $"[{bytes.Length} bytes]"; }
    }
}
